### PR TITLE
Dsprio

### PR DESCRIPTION
**Pull request by ndev2**
Hi @AltraMayor and @cjdoucette ,

Moving from dsprio to skbprio: I've made all the suggested changes requested in the other PR. 

And Cody, just to clear things up I wasn't getting an error when I run the tc class show command in userspace, its just that I didn't see the statistics  change when I passed packets through the scheduler, at least with ping commands. Do I need to be stress testing the qdisc to observe changes?

### CODE REVIEW

File: net/sched/sch_skbprio.c [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+ * queue length limit can be enforced while not restricting the queue lengths
+ * of individual priorities.
+ *
+ * This is especially useful for a denial-of-service defense system; like
+ * SKBprio, which prioritizes packets in flows that demonstrate expected
+ * behavior of legitimate users. The queue is flexible to allow any number
+ * of packets of any priority up to the global limit of the scheduler
+ * without risking resource overconsumption by a flood of low priority packets.
+ *
+ * The SKB Priority Queue standalone codebase is found here:
```
Comment (AltraMayor) [inline net/sched/sch_skbprio.c]: Replace "SKB Priority Queue standalone" with "Gatekeeper".

File: net/sched/sch_skbprio.c [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+#include <linux/types.h>
+#include <linux/kernel.h>
+#include <linux/errno.h>
+#include <linux/skbuff.h>
+#include <net/pkt_sched.h>
+#include <net/sch_generic.h>
+#include <net/inet_ecn.h>
+
+/* Packets are assigned priorities [0, 63] due to the IP DSCP field limits. */
+#define SKBPRIO_MAX_PRIORITY 64
```
Comment (AltraMayor) [inline net/sched/sch_skbprio.c]: Drop the comment and define above since `SKBPRIO_MAX_PRIORITY` is already defined in the header file `net/pkt_sched.h`.

File: include/uapi/linux/pkt_sched.h [hunk 3]
```diff
@@ -124,6 +124,14 @@ struct tc_fifo_qopt {
 	__u32	limit;	/* Queue length: bytes for bfifo, packets for pfifo */
 };
 
+/* SKBPRIO section */
+
+#define SKBPRIO_MAX_PRIORITY 64
```
Comment (AltraMayor) [inline include/uapi/linux/pkt_sched.h]: Add the comment below just before the define of `SKBPRIO_MAX_PRIORITY`.
```
/*
 * Priorities goes from zero to (SKBPRIO_MAX_PRIORITY - 1).
 * SKBPRIO_MAX_PRIORITY should be at least 64 in order for skbprio to be able
 * to map one to one the DS field of IPv4 and IPv6 headers.
 * Memory allocation grows linearly with SKBPRIO_MAX_PRIORITY.
 */
```

File: net/sched/Kconfig [hunk 4]
```diff
@@ -240,6 +240,19 @@ config NET_SCH_MQPRIO
 
 	  If unsure, say N.
 
+config NET_SCH_SKBPRIO
+	tristate "SKB priority queue scheduler (SKBPRIO)"
+	help
+	  Say Y here if you want to use the SKB priority queue
+	  scheduler. This schedules packets according to priorities based on
+	  the DSCP (IPv4) and DS (IPv6) fields, which is useful for request
```
Comment (AltraMayor) [inline net/sched/Kconfig]: Replace "priorities based on the DSCP (IPv4) and DS (IPv6) fields" with "skb->priority"

File: net/sched/Kconfig [hunk 5]
```diff
@@ -240,6 +240,19 @@ config NET_SCH_MQPRIO
...
+	  packets in DoS mitigation systems such as SKBprio.
```
Comment (AltraMayor) [inline net/sched/Kconfig]: Replace "SKBprio" with "Gatekeeper" on the line above.

File: net/sched/sch_skbprio.c [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+
+/*	  SKB Priority Queue
+ *	=================================
+ *
+ * This qdisc schedules a packet according to the value (0-63) of its DSCP
+ * (IPv4) or DS (IPv6) field, where a higher value places the packet closer
```
Comment (AltraMayor) [inline net/sched/sch_skbprio.c]: Replace "the value (0-63) of its DSCP (IPv4) or DS (IPv6) field" with "skb->priority".

File: net/sched/sch_skbprio.c [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+ * to the exit of the queue. Non-IP packets are assigned a default priority
+ * specified to SKBPRIO; if none is specified, default priority is set
+ * to 0. When the queue is full, the lowest priority packet in the queue is
```
Comment (AltraMayor) [inline net/sched/sch_skbprio.c]: Drop "Non-IP packets are assigned a default priority specified to SKBPRIO; if none is specified, default priority is set to 0."

File: net/sched/sch_skbprio.c [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+ * it is dropped.
+ *
+ * Without the SKB priority queue, queue length limits must be imposed
+ * for individual queues, and there is no easy way to enforce a global queue
+ * length limit across all priorities. With the SKBprio queue, a global
+ * queue length limit can be enforced while not restricting the queue lengths
+ * of individual priorities.
+ *
+ * This is especially useful for a denial-of-service defense system; like
+ * SKBprio, which prioritizes packets in flows that demonstrate expected
```
Comment (AltraMayor) [inline net/sched/sch_skbprio.c]: Replace "SKBprio" with "Gatekeeper".

File: net/sched/sch_skbprio.c [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+ * If the packet to be added has lower priority than all packets in the queue,
+ * it is dropped.
+ *
+ * Without the SKB priority queue, queue length limits must be imposed
+ * for individual queues, and there is no easy way to enforce a global queue
+ * length limit across all priorities. With the SKBprio queue, a global
+ * queue length limit can be enforced while not restricting the queue lengths
+ * of individual priorities.
+ *
+ * This is especially useful for a denial-of-service defense system; like
```
Comment (AltraMayor) [inline net/sched/sch_skbprio.c]: Drop ";" after "system".

File: net/sched/sch_skbprio.c [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+		}
+	}
+
+	if (prio > q->highest_prio)
+		q->highest_prio = prio;
+
+	return NET_XMIT_CN;
+drop:
+	qdisc_drop(skb, sch, to_free);
+	return NET_XMIT_SUCCESS | __NET_XMIT_BYPASS;
```
Comment (AltraMayor) [inline net/sched/sch_skbprio.c]: There's no "goto drop" anymore, so we can drop the whole label "drop:".

File: net/sched/sch_skbprio.c [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+			q->lowest_prio = prio;
+
+		sch->q.qlen++;
+		return NET_XMIT_SUCCESS;
+	}
+
+	/* If this packet has the lowest priority, drop it. */
+	lp = q->lowest_prio;
+	if (prio <= lp) {
+		q->qstats[lp].drops++;
```
Comment (AltraMayor) [inline net/sched/sch_skbprio.c]: Replace `lp` with `prio` in the expression `q->qstats[lp].drops++`. So we count the drop of the priority of the packet being dropped instead of the lowest priority available.

File: net/sched/sch_skbprio.c [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+	}
+	return skb;
+}
+
+static int skbprio_change(struct Qdisc *sch, struct nlattr *opt,
+			struct netlink_ext_ack *extack)
+{
+	struct skbprio_sched_data *q = qdisc_priv(sch);
+	struct tc_skbprio_qopt *ctl = nla_data(opt);
+	unsigned int min_limit = 1;
```
Comment (AltraMayor) [inline net/sched/sch_skbprio.c]: Add `const` to the beginning of the line above.

File: net/sched/sch_skbprio.c [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+
+	return 0;
+}
+
+static int skbprio_init(struct Qdisc *sch, struct nlattr *opt,
+			struct netlink_ext_ack *extack)
+{
+	struct skbprio_sched_data *q = qdisc_priv(sch);
+	int prio;
+	unsigned int min_limit = 1;
```
Comment (AltraMayor) [inline net/sched/sch_skbprio.c]: Add `const` to the beginning of the line above.

**Review by AltraMayor — CHANGES_REQUESTED** [review]
**net/sched: add skbprio scheduler**

Notice that among the three statistics that skbprio collects per priority, only the counters of dropped packets is cumulative. That is, unless skbprio is holding packets at the time the statistics are collected, the userspace call will only see zeros.

Either one needs to stress test skbprio to see non-zero counters, or to combine skbprio with another queue discipline that limits bandwidth, so fewer packets are needed to grow the sub-queues and have non-zero counters.

**Comment by ndev2** [comment]
Hi @AltraMayor , I've made the suggested changes. Hope the code is good. Thanks!

File: include/uapi/linux/pkt_sched.h [hunk 14]
```diff
@@ -124,6 +124,21 @@ struct tc_fifo_qopt {
 	__u32	limit;	/* Queue length: bytes for bfifo, packets for pfifo */
 };
 
+/* SKBPRIO section */
+
+/*
+ * Priorities goes from zero to (SKBPRIO_MAX_PRIORITY - 1).
```
Comment (AltraMayor) [inline include/uapi/linux/pkt_sched.h]: `goes` -> `go`

My fault.

**Review by AltraMayor — CHANGES_REQUESTED** [review]
**net/sched: add skbprio scheduler**

You'll need Qiaobin's patch in order to test this new version of our queue discipline. You find his patches here:
kernel patch: https://github.com/mengxiang0811/net-next/pull/2
iproute2 patch: https://github.com/mengxiang0811/iproute2/pull/2

Once you fix the comment I point below, and the tests go fine. You are ready to submit the **second** version of this patch.

**Comment by AltraMayor** [comment]
Also, remember to edit the commit message before pushing the patch upstream. Essentially, you need to adjust the commit message as we've done with the comments in the code.

**Comment by ndev2** [comment]
Hi @AltraMayor  and @cjdoucette, I've stress tested our qdisc with Qiaobin's patch in the kernel  and the tests go fine, the patch is ready to be pushed upstream. However I'm not sure if we should wait since his patch hasn't been applied to the net-next tree yet. How should we proceed?

File: net/sched/sch_skbprio.c [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+	BUG_ON(!to_drop);
+	qdisc_qstats_backlog_dec(sch, to_drop);
+	qdisc_drop(to_drop, sch, to_free);
+
+	q->qstats[lp].backlog -= qdisc_pkt_len(to_drop);
+	q->qstats[lp].drops++;
+
+	__skb_queue_tail(qdisc, skb);
+	qdisc_qstats_backlog_inc(sch, skb);
+	q->qstats[prio].backlog += qdisc_pkt_len(to_drop);
```
Comment (AltraMayor) [inline net/sched/sch_skbprio.c]: Replace the expression `qdisc_pkt_len(to_drop)` with `qdisc_pkt_len(skb)`.

File: net/sched/sch_skbprio.c [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+static int skbprio_change(struct Qdisc *sch, struct nlattr *opt,
+			struct netlink_ext_ack *extack)
+{
+	struct skbprio_sched_data *q = qdisc_priv(sch);
+	struct tc_skbprio_qopt *ctl = nla_data(opt);
+	const unsigned int min_limit = 1;
+
+	if (ctl->limit == (typeof(ctl->limit))-1)
+		q->max_limit = max(qdisc_dev(sch)->tx_queue_len, min_limit);
+	else if (ctl->limit < 1 || ctl->limit > qdisc_dev(sch)->tx_queue_len)
```
Comment (AltraMayor) [inline net/sched/sch_skbprio.c]: Replace `1` with `min_limit`.

**Review by AltraMayor — CHANGES_REQUESTED** [review]
**net/sched: add skbprio scheduler**

Makes the two small changes I'm suggesting here, make sure it compiles without warnings, and do a basic test. You don't need to strees test again.

Once you do the steps above, push the code upstream.

**Comment by ndev2** [comment]
Hi @AltraMayor , we do get a warning when I compiled the code and its due to the min macro:
```
net/sched/sch_skbprio.c: In function ‘skbprio_enqueue’:
./include/linux/kernel.h:810:29: warning: comparison of distinct pointer types lacks a cast
   (!!(sizeof((typeof(x) *)1 == (typeof(y) *)1)))
                             ^
./include/linux/kernel.h:824:4: note: in expansion of macro ‘__typecheck’
   (__typecheck(x, y) && __no_side_effects(x, y))
    ^~~~~~~~~~~
./include/linux/kernel.h:834:24: note: in expansion of macro ‘__safe_cmp’
  __builtin_choose_expr(__safe_cmp(x, y), \
                        ^~~~~~~~~~
./include/linux/kernel.h:843:19: note: in expansion of macro ‘__careful_cmp’
 #define min(x, y) __careful_cmp(x, y, <)
                   ^~~~~~~~~~~~~
net/sched/sch_skbprio.c:101:9: note: in expansion of macro ‘min’
  prio = min(skb->priority, SKBPRIO_MAX_PRIORITY - 1);
         ^~~
```
 Should we make a change here?

But otherwise, everything looks good with the module installation, basic tests, etc, and is ready to be pushed upstream.

**Comment by AltraMayor** [comment]
We must fix this warning:
1. Declare `const unsigned int max_priority = SKBPRIO_MAX_PRIORITY - 1;` in `skbprio_enqueue()`;
2. Update the priority line as follows: `prio = min(skb->priority, max_priority);`.

This should address the warning. Make sure that there's no more warnings or style issues before pushing the code upstream.
